File: sdk-python/pod_protocol/services/escrow.py
"""
Escrow service for managing channel deposits and payments in the PoD Protocol.
"""

import time
from typing import List, Optional, Dict, Any
from dataclasses import dataclass

from .base import BaseService
from ..types import PublicKey, Keypair
from ..utils import find_agent_pda, find_escrow_pda


@dataclass
class DepositEscrowOptions:
    """Options for depositing funds to escrow."""
    channel: PublicKey
    amount: int


@dataclass
class WithdrawEscrowOptions:
    """Options for withdrawing funds from escrow."""
    channel: PublicKey
    amount: int


@dataclass
class EscrowAccount:
    """Escrow account data structure."""
    channel: PublicKey
    depositor: PublicKey
    balance: int
    amount: int  # alias for compatibility
    created_at: int
    last_updated: int
    bump: int


class EscrowService(BaseService):
    """Service for managing channel deposits and payments."""

    async def deposit_escrow(
        self, 
        wallet: Keypair, 
        options: DepositEscrowOptions
    ) -> str:
        """
        Deposit funds to escrow for a channel.
        
        Args:
            wallet: The wallet keypair for signing
            options: Deposit options
            
        Returns:
            Transaction signature
        """
        program = self.ensure_initialized()
        
        # Derive agent PDA
        agent_pda, _ = find_agent_pda(wallet.public_key, self.program_id)
        
        # Derive escrow PDA
        escrow_pda, _ = find_escrow_pda(
            options.channel, 
            wallet.public_key, 
            self.program_id
        )
        
        tx = await program.methods.deposit_escrow(options.amount).accounts({
            "escrow_account": escrow_pda,
            "channel_account": options.channel,
            "depositor_agent": agent_pda,
            "depositor": wallet.public_key,
            "system_program": PublicKey("11111111111111111111111111111112"),
        }).signers([wallet]).rpc(opts={"commitment": self.commitment})
        
        return tx

    async def withdraw_escrow(
        self, 
        wallet: Keypair, 
        options: WithdrawEscrowOptions
    ) -> str:
        """
        Withdraw funds from escrow.
        
        Args:
            wallet: The wallet keypair for signing
            options: Withdraw options
            
        Returns:
            Transaction signature
        """
        program = self.ensure_initialized()
        
        # Derive agent PDA
        agent_pda, _ = find_agent_pda(wallet.public_key, self.program_id)
        
        # Derive escrow PDA
        escrow_pda, _ = find_escrow_pda(
            options.channel, 
            wallet.public_key, 
            self.program_id
        )
        
        tx = await program.methods.withdraw_escrow(options.amount).accounts({
            "escrow_account": escrow_pda,
            "channel_account": options.channel,
            "depositor_agent": agent_pda,
            "depositor": wallet.public_key,
            "system_program": PublicKey("11111111111111111111111111111112"),
        }).signers([wallet]).rpc(opts={"commitment": self.commitment})
        
        return tx

    async def get_escrow(
        self, 
        channel: PublicKey, 
        depositor: PublicKey
    ) -> Optional[EscrowAccount]:
        """
        Get escrow account data.
        
        Args:
            channel: Channel public key
            depositor: Depositor public key
            
        Returns:
            Escrow account data or None if not found
        """
        try:
            escrow_pda, _ = find_escrow_pda(channel, depositor, self.program_id)
            program = self.ensure_initialized()
            account = await program.account.escrow_account.fetch(escrow_pda)
            return self._convert_escrow_account_from_program(account)
        except Exception as e:
            print(f"Escrow not found for channel: {channel}, depositor: {depositor}, error: {e}")
            return None

    async def get_escrows_by_depositor(
        self, 
        depositor: PublicKey, 
        limit: int = 50
    ) -> List[EscrowAccount]:
        """
        Get all escrow accounts by depositor.
        
        Args:
            depositor: Depositor public key
            limit: Maximum number of escrows to return
            
        Returns:
            List of escrow accounts
        """
        try:
            program = self.ensure_initialized()
            filters = [
                {
                    "memcmp": {
                        "offset": 8 + 32,  # After discriminator and channel pubkey
                        "bytes": str(depositor)
                    }
                }
            ]
            
            accounts = await program.account.escrow_account.all(filters=filters)
            
            result = []
            for acc in accounts[:limit]:
                escrow_account = self._convert_escrow_account_from_program(acc.account)
                result.append(escrow_account)
            
            return result
        except Exception as e:
            print(f"Error fetching escrows by depositor: {e}")
            return []

    # Helper Methods
    def _convert_escrow_account_from_program(self, account: Any) -> EscrowAccount:
        """Convert escrow account from program format."""
        balance = account.balance or 0
        return EscrowAccount(
            channel=account.channel,
            depositor=account.depositor,
            balance=balance,
            amount=balance,  # alias for compatibility
            created_at=account.created_at or int(time.time() * 1000),
            last_updated=account.last_updated or int(time.time() * 1000),
            bump=account.bump
        )

File: sdk-js/src/services/escrow.js
/**
 * Escrow service for PoD Protocol SDK
 */

import { BaseService } from './base.js';
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { BN } from '@coral-xyz/anchor';
import { findAgentPDA, findEscrowPDA } from '../utils/pda.js';

/**
 * Service for managing escrow deposits and payments in the PoD Protocol
 * 
 * @class EscrowService
 * @extends BaseService
 */
export class EscrowService extends BaseService {
  /**
   * Deposit funds to escrow for a channel
   * 
   * @param {DepositEscrowOptions} options - Deposit options
   * @param {Keypair} wallet - Depositor's wallet
   * @returns {Promise<string>} Transaction signature
   * 
   * @example
   * ```javascript
   * const tx = await client.escrow.deposit({
   *   channel: channelPDA,
   *   amount: 5000000 // lamports
   * }, wallet);
   * ```
   */
  async deposit(options, wallet) {
    if (!this.isInitialized()) {
      throw new Error('Service not initialized. Call client.initialize() first.');
    }

    // Derive agent PDA
    const [agentPDA] = findAgentPDA(wallet.publicKey, this.programId);

    // Derive escrow PDA
    const [escrowPDA] = findEscrowPDA(options.channel, wallet.publicKey, this.programId);

    return this.retry(async () => {
      const tx = await this.program.methods
        .depositEscrow(new BN(options.amount))
        .accounts({
          escrowAccount: escrowPDA,
          channelAccount: options.channel,
          depositorAgent: agentPDA,
          depositor: wallet.publicKey,
          systemProgram: SystemProgram.programId
        })
        .rpc();

      return tx;
    });
  }

  /**
   * Withdraw funds from escrow
   * 
   * @param {WithdrawEscrowOptions} options - Withdrawal options
   * @param {Keypair} wallet - Depositor's wallet
   * @returns {Promise<string>} Transaction signature
   * 
   * @example
   * ```javascript
   * const tx = await client.escrow.withdraw({
   *   channel: channelPDA,
   *   amount: 1000000 // lamports
   * }, wallet);
   * ```
   */
  async withdraw(options, wallet) {
    if (!this.isInitialized()) {
      throw new Error('Service not initialized. Call client.initialize() first.');
    }

    // Derive agent PDA
    const [agentPDA] = findAgentPDA(wallet.publicKey, this.programId);

    // Derive escrow PDA
    const [escrowPDA] = findEscrowPDA(options.channel, wallet.publicKey, this.programId);

    return this.retry(async () => {
      const tx = await this.program.methods
        .withdrawEscrow(new BN(options.amount))
        .accounts({
          escrowAccount: escrowPDA,
          channelAccount: options.channel,
          depositorAgent: agentPDA,
          depositor: wallet.publicKey,
          systemProgram: SystemProgram.programId
        })
        .rpc();

      return tx;
    });
  }

  /**
   * Get escrow account data
   * 
   * @param {PublicKey} channel - Channel public key
   * @param {PublicKey} depositor - Depositor's public key
   * @returns {Promise<EscrowAccount|null>} Escrow account data
   * 
   * @example
   * ```javascript
   * const escrow = await client.escrow.get(channelPDA, depositorPublicKey);
   * if (escrow) {
   *   console.log('Escrow balance:', escrow.balance);
   * }
   * ```
   */
  async get(channel, depositor) {
    if (!this.isInitialized()) {
      throw new Error('Service not initialized. Call client.initialize() first.');
    }

    try {
      const [escrowPDA] = findEscrowPDA(channel, depositor, this.programId);
      const account = await this.program.account.escrowAccount.fetch(escrowPDA);
      
      return {
        channel: account.channel,
        depositor: account.depositor,
        balance: account.balance?.toNumber() || 0,
        amount: account.balance?.toNumber() || 0, // alias for compatibility
        createdAt: account.createdAt?.toNumber() || Date.now(),
        lastUpdated: account.lastUpdated?.toNumber() || Date.now(),
        bump: account.bump
      };
    } catch (error) {
      if (error.message?.includes('Account does not exist')) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Get all escrow accounts by depositor
   * 
   * @param {PublicKey} depositor - Depositor's public key
   * @param {Object} [options] - Query options
   * @param {number} [options.limit=50] - Maximum number of results
   * @returns {Promise<EscrowAccount[]>} Array of escrow accounts
   * 
   * @example
   * ```javascript
   * const escrows = await client.escrow.getByDepositor(depositorPublicKey, { limit: 20 });
   * ```
   */
  async getByDepositor(depositor, options = {}) {
    if (!this.isInitialized()) {
      throw new Error('Service not initialized. Call client.initialize() first.');
    }

    try {
      const accounts = await this.program.account.escrowAccount.all();
      let escrows = accounts
        .filter(account => account.account.depositor.equals(depositor))
        .map(account => ({
          channel: account.account.channel,
          depositor: account.account.depositor,
          balance: account.account.balance?.toNumber() || 0,
          amount: account.account.balance?.toNumber() || 0,
          createdAt: account.account.createdAt?.toNumber() || Date.now(),
          lastUpdated: account.account.lastUpdated?.toNumber() || Date.now(),
          bump: account.account.bump
        }));

      // Sort by creation date (newest first)
      escrows.sort((a, b) => b.createdAt - a.createdAt);

      if (options.limit) {
        escrows = escrows.slice(0, options.limit);
      }

      return escrows;
    } catch (error) {
      throw new Error(`Failed to get escrows by depositor: ${error.message}`);
    }
  }

  /**
   * Get all escrow accounts for a channel
   * 
   * @param {PublicKey} channel - Channel public key
   * @param {Object} [options] - Query options
   * @param {number} [options.limit=50] - Maximum number of results
   * @returns {Promise<EscrowAccount[]>} Array of escrow accounts
   * 
   * @example
   * ```javascript
   * const escrows = await client.escrow.getByChannel(channelPDA, { limit: 20 });
   * ```
   */
  async getByChannel(channel, options = {}) {
    if (!this.isInitialized()) {
      throw new Error('Service not initialized. Call client.initialize() first.');
    }

    try {
      const accounts = await this.program.account.escrowAccount.all();
      let escrows = accounts
        .filter(account => account.account.channel.equals(channel))
        .map(account => ({
          channel: account.account.channel,
          depositor: account.account.depositor,
          balance: account.account.balance?.toNumber() || 0,
          amount: account.account.balance?.toNumber() || 0,
          createdAt: account.account.createdAt?.toNumber() || Date.now(),
          lastUpdated: account.account.lastUpdated?.toNumber() || Date.now(),
          bump: account.account.bump
        }));

      // Sort by creation date (newest first)
      escrows.sort((a, b) => b.createdAt - a.createdAt);

      if (options.limit) {
        escrows = escrows.slice(0, options.limit);
      }

      return escrows;
    } catch (error) {
      throw new Error(`Failed to get escrows by channel: ${error.message}`);
    }
  }

  /**
   * Get total escrow balance for a channel
   * 
   * @param {PublicKey} channel - Channel public key
   * @returns {Promise<number>} Total escrow balance in lamports
   * 
   * @example
   * ```javascript
   * const totalBalance = await client.escrow.getTotalBalance(channelPDA);
   * console.log('Total escrow balance:', totalBalance);
   * ```
   */
  async getTotalBalance(channel) {
    const escrows = await this.getByChannel(channel);
    return escrows.reduce((total, escrow) => total + escrow.balance, 0);
  }

  /**
   * Check if depositor has sufficient escrow balance for a channel
   * 
   * @param {PublicKey} channel - Channel public key
   * @param {PublicKey} depositor - Depositor's public key
   * @param {number} requiredAmount - Required amount in lamports
   * @returns {Promise<boolean>} True if sufficient balance
   * 
   * @example
   * ```javascript
   * const hasSufficientBalance = await client.escrow.hasSufficientBalance(
   *   channelPDA, 
   *   depositorPublicKey, 
   *   1000000
   * );
   * ```
   */
  async hasSufficientBalance(channel, depositor, requiredAmount) {
    const escrow = await this.get(channel, depositor);
    return escrow ? escrow.balance >= requiredAmount : false;
  }
}
